pysmt/optimization/optimizer.py

```diff
@@ -0,0 +1,703 @@
+        argument or None. If specified, the function will be called
+        each time the algorithm finds a feasible solution. Each call
+        is guaranteed to have a better solution quality than the
+        previous.
+
+        `step_size` the minimum reolution for finding a solution. The
+        optimum will be found in the proximity of `step_size`
+        """
+        rt = None, None
+        if goal.is_maximization_goal() or goal.is_minimization_goal() or goal.is_maxsmt_goal():
+            rt = self._optimize(goal, strategy)
+        else:
+            raise GoalNotSupportedError(self, goal)
+        return rt
+
+    def boxed_optimize(self, goals, strategy='linear'):
+        rt = {}
+        for goal in goals:
+            if goal.is_maximization_goal() or goal.is_minimization_goal() or goal.is_maxsmt_goal():
+                t = self.optimize(goal = goal,strategy = strategy)
+                if t is not None:
+                    rt[goal] = t
+                else:
+                    return None
+            else:
+                raise GoalNotSupportedError(self, goal)
+        return rt
+
+    def lexicographic_optimize(self, goals, strategy='linear'):
+        self._check_pareto_lexicographic_goals(goals, "lexicographic")
+        rt = []
+        client_data = self._setup()
+        for goal in goals:
+            temp = self._lexicographic_opt(client_data, goal, strategy)
+            if temp is None:
+                self._cleanup(client_data)
+                return None
+            else:
+                model, val = temp
+            rt.append(val)
+
+        return model, rt
+
+    def _optimize(self, goal, strategy, extra_assumption = None):
+        """
+        Core algorithm for the optimization process.
+        Based on the strategy, it will either perform a linear or binary search"""
+        _warn_diverge_real_goal(goal)
+        if goal.is_maxsmt_goal():
+            formula = None
+            zero = Real(0) if goal.real_weights() else Int(0)
+            for (c, w) in goal.soft:
+                if formula is not None:
+                    formula = Plus(formula, Ite(c, w, zero))
+                else:
+                    formula = Ite(c, w, zero)
+            assert formula is not None, "Empty MaxSMT goal passed"
+            goal = MaximizationGoal(formula)
+        model = None
+        client_data = self._setup()
+        lower, upper = None, None
+        if not goal.is_maxsmt_goal():
+            term_type = goal.term().get_type()
+            # For bit-vectors, start lower and upper bounds to their width limit
+            if term_type.is_bv_type():
+                # add 1 top upper bound because it is excluded
+                if goal.signed:
+                    lower = -(2**(term_type.width-1))
+                    upper = (2**(term_type.width-1))
+                else:
+                    lower = 0
+                    upper = (2**term_type.width) + 1
```

Comment:
Why put this here and not in the constructor of `OptSearchInterval`?

Reply:
Note after sync: being that minimization has [lower, upper) and maximization has (lower, upper], here we have to check if we are minimizing or maximizing to understand if we have to add 1 to the upper (when maximizing) or remove 1 to the lower (when minimizing)